From: =?utf-8?q?Jonas_=C3=85dahl?= <jadahl@gmail.com>
Date: Thu, 11 Sep 2025 11:05:04 +0200
Subject: wayland/xdg-shell: Ensure applied window geometry is always
 non-empty

Well behaved clients will never end up with a empty window geometry, but
buggy ones could, so to not have to handle that in generic window
management code, try harder to work around broken Wayland client
behavior in Wayland code by be more forgiving when it comes to fix up
incorrectly configured window geometry.

While this will paper over bad client behavior, it'll allow the window
management code remaining ignorent about such awkwardness. Effort is
made to log offending clients to the journal so developers are
potentially made aware of it.

This also makes the window positioning of surfaces with this type of
invalid state more predictable, so update the test case accordingly.

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/4310
Forwarded: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4648
---
 src/tests/wayland-unit-tests.c       |  4 +--
 src/wayland/meta-wayland-xdg-shell.c | 63 +++++++++++++++++++++---------------
 src/wayland/meta-window-wayland.c    |  2 ++
 3 files changed, 41 insertions(+), 28 deletions(-)

diff --git a/src/tests/wayland-unit-tests.c b/src/tests/wayland-unit-tests.c
index beb58bd..241ffe6 100644
--- a/src/tests/wayland-unit-tests.c
+++ b/src/tests/wayland-unit-tests.c
@@ -799,8 +799,8 @@ toplevel_invalid_geometry_basic (void)
   rect = meta_window_config_get_rect (window->config);
   g_assert_cmpint (rect.width, ==, 200);
   g_assert_cmpint (rect.height, ==, 200);
-
-  /* Window position is bogus in this case. */
+  g_assert_cmpint (rect.x, ==, 220);
+  g_assert_cmpint (rect.y, ==, 140);
 
   meta_wayland_test_driver_terminate (test_driver);
   meta_wayland_test_client_finish (wayland_test_client);
diff --git a/src/wayland/meta-wayland-xdg-shell.c b/src/wayland/meta-wayland-xdg-shell.c
index ac3b675..4b609d3 100644
--- a/src/wayland/meta-wayland-xdg-shell.c
+++ b/src/wayland/meta-wayland-xdg-shell.c
@@ -93,6 +93,7 @@ typedef struct _MetaWaylandXdgSurfacePrivate
   struct wl_resource *resource;
   MetaWaylandXdgShellClient *shell_client;
   MtkRectangle geometry;
+  MtkRectangle unconstrained_geometry;
 
   guint configure_sent : 1;
   guint first_buffer_attached : 1;
@@ -1032,9 +1033,10 @@ meta_wayland_xdg_toplevel_post_apply_state (MetaWaylandSurfaceRole  *surface_rol
 
   geometry_changed = !mtk_rectangle_equal (&old_geometry, &window_geometry);
 
-  if (geometry_changed ||
-      pending->derived.surface_size_changed ||
-      pending->has_acked_configure_serial)
+  if (meta_wayland_surface_get_buffer (surface) &&
+      (geometry_changed ||
+       pending->derived.surface_size_changed ||
+       pending->has_acked_configure_serial))
     {
       meta_window_wayland_finish_move_resize (window, window_geometry, pending);
     }
@@ -1986,42 +1988,51 @@ meta_wayland_xdg_surface_post_apply_state (MetaWaylandSurfaceRole  *surface_role
     meta_wayland_xdg_surface_get_instance_private (xdg_surface);
   MetaWaylandShellSurface *shell_surface =
     META_WAYLAND_SHELL_SURFACE (surface_role);
+  MtkRectangle new_geometry = { 0 };
 
   if (pending->has_new_geometry)
+    {
+      priv->unconstrained_geometry = pending->new_geometry;
+      priv->has_set_geometry = TRUE;
+    }
+
+  if (priv->has_set_geometry)
     {
       meta_wayland_shell_surface_determine_geometry (shell_surface,
-                                                     &pending->new_geometry,
-                                                     &priv->geometry);
-      if (priv->geometry.width == 0 || priv->geometry.height == 0)
+                                                     &priv->unconstrained_geometry,
+                                                     &new_geometry);
+
+      if (new_geometry.width == 0 || new_geometry.height == 0)
         {
-          MetaWaylandSurface *surface =
-            meta_wayland_surface_role_get_surface (surface_role);
-          MetaWindow *window = meta_wayland_surface_get_window (surface);
-
-          g_warning ("Client provided invalid window geometry for "
-                     "xdg_surface#%d (%s - %s). Working around.",
-                     wl_resource_get_id (priv->resource),
-                     window ? window->res_class : "N\\A",
-                     window ? window->desc : "N\\A");
-          return;
+          if (pending->has_new_geometry)
+            {
+              MetaWaylandSurface *surface =
+                meta_wayland_surface_role_get_surface (surface_role);
+              MetaWindow *window = meta_wayland_surface_get_window (surface);
+
+              g_warning ("Client provided invalid window geometry for "
+                         "xdg_surface#%d (%s - %s). Working around.",
+                         wl_resource_get_id (priv->resource),
+                         window ? window->res_class : "N\\A",
+                         window ? window->desc : "N\\A");
+            }
+          meta_wayland_shell_surface_calculate_geometry (shell_surface,
+                                                         &new_geometry);
         }
-
-      priv->has_set_geometry = TRUE;
     }
-  else if (!priv->has_set_geometry)
+  else
     {
-      MtkRectangle new_geometry = { 0 };
-
       /* If the surface has never set any geometry, calculate
        * a default one unioning the surface and all subsurfaces together. */
 
       meta_wayland_shell_surface_calculate_geometry (shell_surface,
                                                      &new_geometry);
-      if (!mtk_rectangle_equal (&new_geometry, &priv->geometry))
-        {
-          pending->has_new_geometry = TRUE;
-          priv->geometry = new_geometry;
-        }
+    }
+
+  if (!mtk_rectangle_equal (&new_geometry, &priv->geometry))
+    {
+      pending->has_new_geometry = TRUE;
+      priv->geometry = new_geometry;
     }
 }
 
diff --git a/src/wayland/meta-window-wayland.c b/src/wayland/meta-window-wayland.c
index d6a58f7..9525af3 100644
--- a/src/wayland/meta-window-wayland.c
+++ b/src/wayland/meta-window-wayland.c
@@ -1336,6 +1336,8 @@ meta_window_wayland_finish_move_resize (MetaWindow              *window,
   MetaWaylandToplevelDrag *toplevel_drag;
   MetaPlaceFlag place_flags = META_PLACE_FLAG_NONE;
 
+  g_return_if_fail (!mtk_rectangle_is_empty (&new_geom));
+
   /* new_geom is in the logical pixel coordinate space, but MetaWindow wants its
    * rects to represent what in turn will end up on the stage, i.e. we need to
    * scale new_geom to physical pixels given what buffer scale and texture scale
